compiler/src/lnntoamm/Expr.ts

```diff
@@ -0,0 +1,279 @@
+    let asts = ast.getAll().map(a => a.get('baseassignable'));
+    let generated = [];
+    let expr: Expr = null;
+    for (let ii = 0; ii < asts.length; ii++) {
+      let work = asts[ii];
+      if (work.has('objectliterals')) {
+        TODO('object literals');
+      } else if (work.has('functions')) {
+        TODO('functions in functions');
+      } else if (work.has('variable')) {
+        const varName = work.get('variable').t;
+        if (ii === asts.length - 1) {
+          let dec = metadata.get(varName);
+          if (dec === null) {
+            throw new Error(`${varName} not defined`);
+          }
+          expr = dec.ref();
+          break;
+        }
+        const next = asts[ii + 1];
+        if (next.has('fncall')) {
+          let text = `${expr !== null ? expr.ast.t.trim() + '.' : ''}${varName}${next.get('fncall').t.trim()}`;
+          let and: any = {
+            fnname: work.get('variable'),
+            fncall: next.get('fncall'),
+          };
+          let accessed: Ref | null = null;
+          // DO NOT access `expr` past this block until it is set.
+          if (expr !== null) {
+            and.fnaccess = expr.ast;
+            let dec = Dec.gen(expr, metadata);
+            generated.push(dec);
+            accessed = dec.ref();
+            expr = null;
+          }
+          let callAst = new NamedAnd(
+            text,
+            and,
+            (work as NamedOr).filename,
+            work.line,
+            work.char,
+          );
+          let [intermediates, call] = Call.fromCallAst(
+            callAst,
+            varName,
+            accessed,
+            metadata,
+          );
+          generated.push(...intermediates);
+          expr = call;
+          ii += 1;
+        } else if (next.has('methodsep')) {
+          TODO('accesses/methods on non-constants');
+        } else {
+          throw new Error(`unexpected token: expected dot or call, found ${next.t.trim()}`);
+        }
+      } else if (work.has('constants')) {
+        work = work.get('constants');
+        if (expr !== null) {
+          throw new Error(`unexpected constant found`);
+        }
+        let [int, constant] = Const.fromConstantsAst(work, metadata);
+        generated.push(...int);
+        expr = constant;
+      } else {
+        // TODO: don't lump in HOF and chains
+        throw new Error(`unexpected token: expected variable or value, found ${work.t.trim()}`);
+      }
+    }
+    return [generated, expr];
+  }
+
+  static fromAssignablesAst(ast: LPNode, metadata: MetaData): [Stmt[], Expr] {
+    // break it up so that we're only working on one base assignable list or operator at a time.
+    let operated = ast.getAll().map(work => {
+      work = work.get('withoperators');
+      if (work.has('baseassignablelist')) {
+        return Expr.fromBaseassignablelist(work.get('baseassignablelist'), metadata);
+      } else if (work.has('operators')) {
+        return work;
```

Comment:
Should there be a TODO here, too? I don't see how this branch can be right?

Reply:
yes, this should be a TODO